src/app/guards/auth.guard.ts

```diff
@@ -8,11 +8,41 @@ import { KeycloakService } from 'app/services/keycloak.service';
 export class AuthGuard implements CanActivate {
   constructor(private readonly keycloakService: KeycloakService, private readonly router: Router) {}
 
-  canActivate(): true | UrlTree {
-    if (this.keycloakService.isAuthenticated()) {
-      return true;
+  canActivate(): boolean | UrlTree {
+    // When a successful login occurs, we store the identity provider used in sessionStorage.
+    const lastIdp = sessionStorage.getItem(this.keycloakService.LAST_IDP_AUTHENTICATED);
+
+    // Not authenticated
+    if (!this.keycloakService.isAuthenticated()) {
+      if (lastIdp === null) {
+        // If an identity provider hasn't been selected then show the login page.
+        return this.router.parseUrl('/login');
+      }
+      // If an identity provider was already selected and successfully authenticated
+      // then do a keycloak login with that identity provider.
+      this.keycloakService.login(lastIdp);
+      return false;
+    }
+
+    if (lastIdp === null) {
+      // Store the identity provider that was used to successfully log in.
+      // Even if the user is unauthorized, we still want to store this because
+      // we don't have a logout, so there is no point allowing the user to select
+      // a different IDP, as Keycloak will just ignore the selection when the user
+      // is authenticated already.
+      const idp = this.keycloakService.getIdpFromToken();
+      if (idp !== '') {
+        sessionStorage.setItem(this.keycloakService.LAST_IDP_AUTHENTICATED, idp);
+      }
```

Comment:
There should be a `sessionStorage.removeItem(this.keycloakService.LAST_IDP_AUTHENTICATED)` if there is no IDP.

Reply:
The only time there would be no IDP is when there is no token.  In that case, `this.keycloakService.isAuthenticated()` would be false, and it would never reach this code.  
```
    if (!this.keycloakService.isAuthenticated()) {
      if (lastIdp === null) {
        // If an identity provider hasn't been selected then show the login page.
        return this.router.parseUrl('/login');
      }
      // If an identity provider was already selected and successfully authenticated
      // then do a keycloak login with that identity provider.
      this.keycloakService.login(lastIdp);
      return false;
    }
```